ratis-examples/src/main/java/org/apache/ratis/examples/filestore/FileStoreStateMachine.java

```diff
@@ -153,6 +163,55 @@ public TransactionContext startTransaction(RaftClientRequest request) throws IOE
     return reply.thenApply(ExamplesProtos.ReadReplyProto::getData);
   }
 
+  class LocalStream implements DataStream {
+    private final DataChannel dataChannel;
+
+    LocalStream(DataChannel dataChannel) {
+      this.dataChannel = dataChannel;
+    }
+
+    @Override
+    public DataChannel getDataChannel() {
+      return dataChannel;
+    }
+
+    @Override
+    public CompletableFuture<?> cleanUp() {
+      return CompletableFuture.supplyAsync(() -> {
+        try {
+          dataChannel.close();
+          return true;
+        } catch (IOException e) {
+          return FileStoreCommon.completeExceptionally("Failed to close data channel", e);
+        }
+      });
+    }
+  }
+
+  @Override
+  public CompletableFuture<DataStream> stream(RaftClientRequest request) {
+    final ByteString reqByteString = request.getMessage().getContent();
+    final StreamWriteRequestHeaderProto proto;
+    try {
+      proto = StreamWriteRequestHeaderProto.parseFrom(reqByteString);
+    } catch (InvalidProtocolBufferException e) {
+      return FileStoreCommon.completeExceptionally(
+          "Failed to parse stream header", e);
+    }
+    return files.createDataChannel(proto.getPath().toStringUtf8())
+        .thenApply(LocalStream::new);
+  }
+
+  @Override
+  public CompletableFuture<?> link(DataStream stream, LogEntryProto entry) {
+    LOG.info("link {}", stream);
+    if (stream == null) {
+      return JavaUtils.completeExceptionally(new IllegalStateException("Null stream: entry=" + entry));
+    }
+    dataStreamToLogEntryMap.put(stream, entry);
```

Comment:
My current understanding is link is just to preserve the mapping between DataStream and LogEntryProto.

Reply:
In applyTransaction(..), a log entry will be given and then the state machine has to apply it.

For Streaming in FileStore, we may simply check if the stream is closed and the file exists since the data should have already been written to the file.

Reply:
link(..) is to tell the state machine which stream to use in applyTransaction(..).  In FileStore, the request header has the path so that we may use it to look up the file.  The path is the key to get the corresponding data.   link(..) in FileStore may not be very useful since we have the path in the header.

Suppose we don't have the path in the header.  Then link(..) becomes very important since it can map LogEntry to data.  Otherwise, we don't know which file should be used in applyTransaction(..).

Reply:
I see. 

And also probably for `write`, for the streaming case it should does a no-op (it was implemented by this way already which is nice)